File: frontend/types/scan.ts
export type RiskLabel = "HIGH" | "MEDIUM" | "LOW"
export type ScanStatus = "pending" | "processing" | "completed" | "failed"
export type Ecosystem = "npm" | "python" | "go" | "rust" | "java" | "ruby" | "php"

export interface ScanSummary {
  total_packages: number
  high_risk: number
  medium_risk: number
  low_risk: number
  avg_bus_factor: number
  avg_scorecard: number
}

export interface Scan {
  id: string
  filename: string
  ecosystem: Ecosystem
  status: ScanStatus
  created_at: string
  completed_at: string | null
  summary: ScanSummary
  package_names: string[]
}

export interface PackageMaintainer {
  login: string
  commits_90d: number
  last_commit: string | null
}

export interface PackageRisk {
  name: string
  ecosystem: string
  weekly_downloads: number
  bus_factor: number
  risk_score: number
  risk_label: RiskLabel
  last_release_months_ago: number | null
  last_commit_at: string | null
  open_issues: number | null
  scorecard_score: number | null
  maintainers: PackageMaintainer[]
  hops_from_root: number | null
}

export interface GraphNode {
  id: string
  labels: string[]
  properties: Record<string, unknown>
}

export interface GraphRelationship {
  id: string
  type: string
  startNodeId: string
  endNodeId: string
  properties: Record<string, unknown>
}

export interface GraphData {
  nodes: GraphNode[]
  relationships: GraphRelationship[]
}

File: backend/app/schemas/scan.py
from datetime import datetime
from typing import Literal

from pydantic import BaseModel


class ScanSummaryResponse(BaseModel):
    total_packages: int
    high_risk: int
    medium_risk: int
    low_risk: int
    avg_bus_factor: float
    avg_scorecard: float


class ScanResponse(BaseModel):
    id: str
    filename: str
    ecosystem: str
    status: Literal["pending", "processing", "completed", "failed"]
    created_at: datetime
    completed_at: datetime | None
    summary: ScanSummaryResponse
    package_names: list[str]


class PackageRisk(BaseModel):
    name: str
    ecosystem: str
    weekly_downloads: int
    bus_factor: int
    risk_score: float
    risk_label: Literal["HIGH", "MEDIUM", "LOW"]
    last_release_months_ago: float | None
    last_commit_at: str | None
    open_issues: int | None
    scorecard_score: float | None
    maintainers: list[dict]
    hops_from_root: int | None = None
